### PR TITLE
SMT names that match Lean names

### PR DESCRIPTION
**Pull request by dranov**
This is a **draft** PR addressing #7. I am still working on this, but want some feedback early.

As of now, sorts and terms in SMT-LIB get names that are derived from the Lean names. There are a few remaining hiccups:

- [ ] `etom`s are still named `stmi_<N>`
- [ ] inductive data types have constructors named `cifvar_<N>`– I'm not sure what is going on there
- [x] disposable names, used for quantifier binders, still have uninformative names, e.g. `stmd_80`

See the test case in `Test/SmtTranslation/Names.lean` for a large example.

----

As part of this PR, I also want to give the `valid_fact_<N>` assertions (part of unsat cores) potentially more user-readable names, matching the hypothesis in the Lean context where `auto` is called. That seems more straightforward.

### CODE REVIEW

**Comment by dranov** [comment]
@PratherConid I'm not sure where to get `etom` names. What do they correspond to on the Lean side? Could you please point me in the right direction?

File: Auto/Tactic.lean [hunk 1]
```diff
@@ -284,13 +284,18 @@ def queryTPTP (exportFacts : Array REntry) : LamReif.ReifM (Array Embedding.Lam.
 
 open Embedding.Lam in
 def querySMT (exportFacts : Array REntry) (exportInds : Array MutualIndInfo) : LamReif.ReifM (Option Expr) := do
-  let lamVarTy := (← LamReif.getVarVal).map Prod.snd
+  -- GEORGE: do we need to pass more of `LamReif:State` to `lamFOL2SMT`?
+  let lamVarTy ← LamReif.getVarVal
+  trace[auto.lamReif.printValuation] "lamVarTy: {lamVarTy}"
   let lamEVarTy ← LamReif.getLamEVarTy
+  trace[auto.lamReif.printValuation] "lamEVarTy: {lamEVarTy}"
```
Comment (dranov) [inline Auto/Tactic.lean:291]: One thing I'm confused about is also that `printValuation` (in `LamReif.lean`) doesn't print any `Etom` lines (suggesting `lamEVarTy` is empty), but this prints `lamEVarTy: [#0, #1, #1, #2]` on the `Names.lean` example.

Is `printValuation` called too early?

File: Auto/Translation/Assumptions.lean [hunk 2]
```diff
@@ -10,16 +10,55 @@ open Lean
...
+  /-- Inhabitation instance synthesized for canonicalized type -/
+  | tyCanInh : Origin
+  /-- Inductive type recursor -/
+  | recursor : Name → Origin
+  /-- Rewrite -/
+  | rewrite : Origin
+  /-- Lemma proved by native prover -/
+  | byNative : Name → Origin
+  /-- Catch-all for string-based origins -/
+  | generic : String → Origin
```
Comment (dranov) [inline Auto/Translation/Assumptions.lean:36]: I've begun to add some structure to the stringly-typed proof trees. The goal is to eventually extract a user-readable lemma name from the `DTr` of assertions in the unsat core.

File: Auto/Tactic.lean [hunk 3]
```diff
@@ -284,13 +284,18 @@ def queryTPTP (exportFacts : Array REntry) : LamReif.ReifM (Array Embedding.Lam.
...
+  let tyVal ← LamReif.getTyVal
+  trace[auto.lamReif.printValuation] "tyVal: {tyVal}"
   let exportLamTerms ← exportFacts.mapM (fun re => do
     match re with
     | .valid [] t => return t
     | _ => throwError "runAuto :: Unexpected error")
-  let commands ← (lamFOL2SMT lamVarTy lamEVarTy exportLamTerms exportInds).run'
+  let commands ← (lamFOL2SMT lamVarTy lamEVarTy tyVal exportLamTerms exportInds).run'
```
Comment (dranov) [inline Auto/Tactic.lean:298]: We're threading a lot of the `LamReif` `State` into `lamFOL2SMT` and associated functions. Would it be cleaner to make this part of relevant monad?

If yes, could you suggest how to do that? I haven't worked with monads in Lean before.

File: Auto/Tactic.lean (hunk 1 above)
Comment (PratherConid) [reply Auto/Tactic.lean:291]: ``etom``s are created by skolemization. Maybe we can use names like ``sk_<num>``.

File: Auto/Tactic.lean (hunk 3 above)
Comment (PratherConid) [reply Auto/Tactic.lean:298]: Your approach looks good to me. Using a monad might make things cleaner, but I'm not sure. Anyways, I'll clean up the code when you're done, so you can keep going for now.

File: Auto/Tactic.lean (hunk 1 above)
Comment (PratherConid) [reply Auto/Tactic.lean:291]: Some automated theorem provers use ``sk`` for skolems, but other names are also ok.

**Comment by dranov** [comment]
With regards to `disposableName` (used in `lamQuantified2STerm`), is the original Lean name of the binder kept anywhere?

**Comment by PratherConid** [comment]
No, Lean-auto does not keep the name of the original binder. I did not include it in ``LamTerm`` because it might affect the performance of the checker.

**Comment by PratherConid** [comment]
I think you can also ignore this for now. Supporting SMT binder names that correspond to Lean binder names will require systematic changes to the codebase. An easier approach would be giving binders names like `a`, `b`, `c`.

**Comment by dranov** [comment]
Alright, that's fair. For now I've settled on taking the binder names from the associated sorts, so for instance you now get `n<idx>` for a binder of sort `node` rather than `smtd_<idx>`.

**Comment by dranov** [comment]
I've removed the unrelated commit making `DTr` less stringly-typed.

**Comment by dranov** [comment]
@PratherConid I guess this is good enough for now. Could you please take a look, clean-up as necessary, and merge?

File: Auto/IR/SMT.lean [hunk 4]
```diff
@@ -415,29 +426,66 @@ section
...
+      -- to disambiguate, we use the argument names, not just the function names.
+      | Expr.app f args =>
+        let fname ← smtNameFromHint f default
+        let argname ← smtNameFromHint args "arg"
+        return s!"{fname}_{argname}"
+      | Expr.proj typeName idx _ => return s!"{typeName.toString}_proj_{idx}"
+      | _ =>
+        trace[auto.lamReif.printValuation] "smtNameFromHint :: Unexpected hint; saw {nameHint.ctorName} {nameHint} ({nameHint.dbgToString})"
+        return default
+
```
Comment (PratherConid) [inline Auto/IR/SMT.lean:469]: The above ``smtNameFromHint`` will print out the entire expression, including all implicit arguments, which will result in very long strings in many cases. Consider the following example:
```
section

  variable (A : Type) (inst : Add A)

  axiom Rat_comm (a b : Rat) : a + b = b + a

  theorem Rat_comm' (a b : Rat) : a + b = b + a := by
    auto [Rat_comm]

end
```
What the pretty printer prints as ``@HAdd.hAdd Rat Rat Rat instHAdd a b`` (not showing implicit arguments) becomes ``HAdd.hAdd_Lean.Rat_Lean.Rat_Lean.Rat_instHAdd_Lean.Rat_Lean.Rat.instAddRat``. This issue will be more serious for ``MathLib4`` problems. For example, when using ``MathLib4``, what prints as ``⇑h`` can mean
```
@DFunLike.coe.{u_2 + 1, u_1 + 1, max (u_1 + 1) (u_2 + 1)}
    (@AddMonoidHom.{u_2, u_1} A₀ A₁
      (@AddMonoid.toAddZeroClass.{u_1} A₀
        (@SubNegMonoid.toAddMonoid.{u_1} A₀
          (@AddGroup.toSubNegMonoid.{u_1} A₀ (@AddCommGroup.toAddGroup.{u_1} A₀ inst✝⁵))))
      (@AddMonoid.toAddZeroClass.{u_2} A₁
        (@SubNegMonoid.toAddMonoid.{u_2} A₁
          (@AddGroup.toSubNegMonoid.{u_2} A₁ (@AddCommGroup.toAddGroup.{u_2} A₁ inst✝²)))))
    A₀ (fun (x : A₀) => A₁)
    (@AddMonoidHom.instFunLike.{u_2, u_1} A₀ A₁
      (@AddMonoid.toAddZeroClass.{u_1} A₀
        (@SubNegMonoid.toAddMonoid.{u_1} A₀
          (@AddGroup.toSubNegMonoid.{u_1} A₀ (@AddCommGroup.toAddGroup.{u_1} A₀ inst✝⁵))))
      (@AddMonoid.toAddZeroClass.{u_2} A₁
        (@SubNegMonoid.toAddMonoid.{u_2} A₁
          (@AddGroup.toSubNegMonoid.{u_2} A₁ (@AddCommGroup.toAddGroup.{u_2} A₁ inst✝²)))))
    h
```
In short, we might want to keep the smt Name closer to what the pretty printer prints. I think ``← e.toMessageData.format`` will do the job. Maybe append index when necessary to avoid collision.

Comment (PratherConid) [reply Auto/IR/SMT.lean:469]: and replace non-unicode characters with underscore

File: Auto/IR/SMT.lean [hunk 5]
```diff
@@ -415,29 +426,66 @@ section
...
     let h2lMap ← getH2lMap
     if let .some name := h2lMap.find? cstr then
       return name
     let idx ← getIdx
-    let currName : String := s!"smti_{idx}"
-    if l2hMap.contains currName then
-      throwError "h2Symb :: Unexpected error"
+    let defaultName : String := s!"smti_{idx}"
+    let mut currName ← smtNameFromHint nameHint defaultName
+    trace[auto.lamReif.printValuation] "smti_{idx} := {currName} (from expr: {nameHint} | lam: {cstr})"
```
Comment (PratherConid) [inline Auto/IR/SMT.lean:483]: ``auto.lamReif.printValuation`` is inappropriate. May be register a new trace class, e.g. ``auto.smt.h2symb``

File: Auto/IR/SMT.lean (hunk 4 above)
Comment (PratherConid) [reply Auto/IR/SMT.lean:469]: Sorry. To pretty print, use ``(← PrettyPrinter.delab e).raw.prettyPrint``.

Comment (PratherConid) [reply Auto/IR/SMT.lean:469]: ``← PrettyPrinter.formatTerm (← PrettyPrinter.delab e).raw`` might be even better

File: Auto/Translation/LamFOL2SMT.lean [hunk 6]
```diff
@@ -429,21 +453,23 @@ def termAuxDecls : Array IR.SMT.Command :=
   `valid_fact_{i}` corresponds to the `i`-th entry in `facts`
 -/
 def lamFOL2SMT
-  (lamVarTy lamEVarTy : Array LamSort)
+  (lamVarTy : Array (Expr × LamSort))
```
Comment (PratherConid) [inline Auto/Translation/LamFOL2SMT.lean:456]: Inappropriate naming across entire file! Should be named ``varVal`` because it contains the valuation of the term.

Comment (PratherConid) [reply Auto/Translation/LamFOL2SMT.lean:456]: Valuation for term atoms should be eta-reduced. This is likely an issue irrelevent to this pull request.

**Comment by PratherConid** [comment]
``← PrettyPrinter.formatTerm (← PrettyPrinter.delab e).raw`` might also be used to pretty-print ``DTr``.

File: Auto/IR/SMT.lean [hunk 7]
```diff
@@ -277,6 +280,17 @@ def SMTOption.toString : SMTOption → String
...
+def SMTReservedWords : HashSet String :=
+  let reserved := #[ "as", "let", "exists", "forall", "match", "par",
+    "assert", "check-sat", "check-sat-assuming",
+    "declare-const", "declare-datatype", "declare-datatypes",
+    "declare-fun", "declare-sort", "define-fun", "define-fun-rec", "define-funs-rec",
+    "define-sort", "echo", "exit", "get-assertions", "get-info",
+    "get-model", "get-option", "get-proof", "get-unsat-assumptions",
+    "get-unsat-core", "get-value", "pop", "push", "reset", "reset-assertions",
+    "set-info", "set-logic", "set-option"]
+  reserved.foldl (fun hs s => hs.insert s) {}
```
Comment (PratherConid) [inline Auto/IR/SMT.lean:292]: Not sufficient. Pre-defined symbols in SMT-lib like ``+, -, *, /, %, ...`` and auxiliary definitions in Lean-auto like `iemod` should be excluded. Individual SMT solvers also have pre-defined symbols.

Comment (PratherConid) [reply Auto/IR/SMT.lean:292]: Already causing issue in
``
example : (∀ (xs ys zs : List α), xs ++ ys ++ zs = xs ++ (ys ++ zs)) := by
  auto [List.append_assoc]
``
after previous pretty-print commit.

Comment (PratherConid) [reply Auto/IR/SMT.lean:292]: image

Comment (PratherConid) [reply Auto/IR/SMT.lean:292]: @dranov The reserved words also contain numeral, decimal and string, but there is no guarantee that Lean's pretty printer won't print them. I'm feeling that using absolutely no predetermined prefix (previously ``smti_``) for Lean-auto generated names is causing too much trouble. Can we maybe add something like ``_`` as a prefix?

**Comment by PratherConid** [comment]
I'll merge the pull request for now.

**Comment by PratherConid** [comment]
Made some significant changes in ff56bdbb2a800edea572933eb800ad913cf2ba74, behaviour on ``SMTTranslation/Names.lean`` becomes a bit different.
